### PR TITLE
[batch] Add attempts with format version < 3 into attempt_resources

### PR DESCRIPTION
**Pull request by jigold**
cc: @danking 

Here is the first major database migration. The goal is to add all attempt resources into the database for any attempt with batch format version less than 3. The node configuration was the same for all attempts back then with standard instances, 16 cores, and a 100GB boot disk. I figured out what the quantity for each resource should be by looking at how we compute quantities for resources in `batch/batch/resources.py`. I checked the inserted quantities are identical to the attempts that already exist in the database right after we converted the billing over to using resources.

Once we have all resources for all attempts, the next step (future PR) is to do a scan and repopulate the new aggregated billing tables **by date**. In this PR, I don't try and add the usage to the existing `aggregated_*_resources` tables. I did this to cut down on time and space since we're eventually going to deprecate those tables anyways. Because I don't touch those tables, we don't need to worry about modifying the client code and how the current billing information is calculated.

How this migration works is there are 5 phases:
1. Compute the expected number of attempts to process for format version < 3. 
2. Divide the search space into chunks of size 100 attempts (empirically determined this was the best chunk size) and randomize the order of the chunks.
3. Serially process 5000 chunks with only 10 out of the 100 records as a "burn in period" to avoid the birthday problem when trying to insert records in parallel.
4. In parallel, process all the chunks with 10 way parallelism (empirically determined to max CPU for a 4 core db instance)
5. Do an audit of the results to make sure the attempt resources now has the correct number of rows and the billing is within $0.001 per job with the old way and new way of computing the billing. The tolerance of $0.001 was empirically determined. At a threshold of $0.0001, 33/30,000,000 attempts failed. I think this is good enough as there's always going to be rounding errors.

I did not do an explicit audit in the code to make sure the other aggregated_*_resources tables did not change. I spot checked this was correct in my test database. To do the complete audit during the actual migration would take more time.

I made sure all the inserts were idempotent. Please double check this.

The inserts use a temporary table with an isolation level of read committed. The reason for this is because `INSERT INTO ... SELECT` locks the next gap lock if the isolation level is not read committed. Maybe what I did is overkill and it's no longer a problem with the new burn in period. https://dev.mysql.com/doc/refman/8.0/en/innodb-locks-set.html I'd be interested to hear @danking feedback on what the best query here is to allow parallelism.

There are ~30 million attempts that need to be processed for hail-vdc (~60% of the attempts). This will add ~20Gi to the existing database. I use 4 cores to get this migration to be ~3 hours, so we will want to **upgrade the database to 8 cores** while this migration is running. The inserting takes about 2 hours and the audit is 45 minutes or so.

For the Australians, I think this script should be a no-op because I believe they started their instance after the billing changes went in around May/June 2020.

The main thing to look out for is whether I got the interval queries correct! For example:

```python3
                where_cond = 'WHERE (attempts.batch_id > %s OR ' \
                             '(attempts.batch_id = %s AND attempts.job_id > %s) OR ' \
                             '(attempts.batch_id = %s AND attempts.job_id = %s AND attempts.attempt_id >= %s)) ' \
                             'AND (attempts.batch_id < %s OR ' \
                             '(attempts.batch_id = %s AND attempts.job_id < %s) OR ' \
                             '(attempts.batch_id = %s AND attempts.job_id = %s AND attempts.attempt_id < %s))'
```

The last thing is to do a sanity check and triple check the new database trigger won't run anything for format version < 3. Also, a sanity check that format_version < 3 is the right cutoff. I got this by looking at the `BatchFormatVersion` class where we have the cost function depending on format version.

### CODE REVIEW

**Comment by jigold** [comment]
This is fine to review, but before it merges, I need to slightly modify a query so it's compatible with future versions of MySQL.

**Comment by jigold** [comment]
Also, I forgot to test that the query that does the audit doesn't lock any tables from writes.

File: batch/sql/insert_attempt_resources_format_version_lt_3.py
Comment (daniel-goldstein) [inline batch/sql/insert_attempt_resources_format_version_lt_3.py:136]: This doesn't really matter but could just be `LIMIT 1`. 100 feels weirdly arbitrary.

**Review by daniel-goldstein — CHANGES_REQUESTED** [review]
I think this looks good. Just a few small questions.

File: batch/sql/insert_attempt_resources_format_version_lt_3.py
Comment (jigold) [reply batch/sql/insert_attempt_resources_format_version_lt_3.py:58]: I can remove it. I used it for testing what the speed would be for inserting single variants.

File: batch/sql/insert_attempt_resources_format_version_lt_3.py
Comment (jigold) [reply batch/sql/insert_attempt_resources_format_version_lt_3.py:84]: Nope. Good catch!

**Comment by jigold** [comment]
I need to test out the last change I made to fix a potential forward compatibility issue when we do one last final test. But for now, I think it would be good to keep reviewing further up the stack to make sure we are happy with all of the changes before I do one last test.

**Comment by jigold** [comment]
This is working now. ~2.5 hours and adds 35 GB of storage. The jobs table needs to be in the join because we use `cores_mcpu` in the insert statement.

File: batch/sql/insert_attempt_resources_format_version_lt_3.py
Comment (jigold) [reply batch/sql/insert_attempt_resources_format_version_lt_3.py:82]: Yes we do need it because of the `cores_mcpu` that we're using in the quantity calculation

**Comment by jigold** [comment]
The test migration worked! There's still two more PRs that need to go in before this one, but otherwise it's all set.

**Comment by jigold** [comment]
==========================================
**Do NOT merge without setting the database to 8 cores
==========================================

**Comment by jigold** [comment]
@daniel-goldstein Should be all set. Let's plan on merging this early on Friday morning.

**Review by daniel-goldstein — APPROVED** [review]

**Pull request merged**
